### ISSUE TITLE
Accessing only unchanged values of a snapshot would prevent re-renders

### ISSUE THREAD

**Issue by timuric** [pmndrs/valtio#209]
Normally extracting full snapshot would broadcast all changes, however if only unchanged values are being accessed it will prevent re-rendering.
Try to uncomment line 14 to see the desired behaviour

https://codesandbox.io/s/valtio-444r4?file=/src/App.tsx:0-509

```js
import { proxy, useSnapshot } from "valtio";

class State {
  count = 0;
  value = 0;
}

const store = proxy(new State());

export default function App() {
  const snapshot = useSnapshot(store);

  //Comment this line out to enable rerender
  const value = snapshot.value;

  return (
    <>
      {/* Try adding values */}
      <button onClick={() => (store.count += 1)}>add</button>
      {/* Will show stale value */}
      <button onClick={() => alert(snapshot.count)}>show value</button>
    </>
  );
}
```

**Comment by dai-shi** [comment]
Hi, thanks for trying out.

snapshot is for using only in render. Because onClick handler is outside render, it's recommended to use store instead of snapshot.
```jsx
<button onClick={() => alert(store.count)}>show value</button>
```

We have [eslint-plugin-valtio](http://github.com/pmndrs/eslint-plugin-valtio) to detect this misusage.

---

The reason commenting out changes the behavior is a different story. If no property access happens during render, the snapshot is assumed to be used as a whole.

Hope it helps clarifying.

**Comment by timuric** [comment]
@dai-shi thank you so much for your quick reply and helpful response! I would definitely use the es-plugin from now on.

I was under the impression that ```const snapshot = useSnapshot(store)``` would always trigger re-render. Perhaps would be nice to add some kind of gotchas section to readme also a link to es-plugin would be useful. Let me know if you would like a pull request.

**Comment by dai-shi** [comment]
I was planning to add eslint-plugin-valtio in README, once https://github.com/pmndrs/eslint-plugin-valtio/issues/4 is resolved. (@Aslemammad are you available?)

Meanwhile, let me explain why it's a gotcha, but not wrong. (and when you get comfortable, a PR would be great.)

Suppose we have this state (or store).
```js
const state = proxy({
  obj: {
    count: 0,
    text: 'hello',
  },
})
```

If using the snapshot with accessing count,
```js
const snap = useSnapshot(state)
snap.obj.count
```

it will re-render only if `count` changes.

If the property access is obj,
```js
const snap = useSnapshot(state)
snap.obj
```

then, it will re-render if `obj` changes. This includes `count` changes and `text` changes.

Now, we can subscribe to the portion of the state.
```js
const snapObj = useSnapshot(state.obj)
snapObj
```

This is technically same as the previous one. It doesn't touch the property of `snapObj`, so it will re-render if `obj` changes.

In summary, if a snapshot object (nested or not) is not accessed with any properties, it assumes the entire object is accessed, so any change inside the object will trigger re-render.

**Comment by timuric** [comment]
@dai-shi brilliant explanation, thank you! 

I think your comment as it is, would be a great addition to the documentation.

Btw the very first thing I did before started using Valtio in my project was testing how would re-render work with these two cases 
```js
const snap = useSnapshot(state)
```
```js
const {count} = useSnapshot(state)
```
and in my head it sounded like: "Great! I can subscribe to all changes or only those that I need" 
But now I realized that re-render optimization is even more advance, which is a great selling point!

Referencing store in handlers actually makes a lot of sense, maybe worth pointing out that using "snapshots" in handlers is an anti-pattern. 

Just want to say that you did great work with Valtio! It's refreshingly simple and powerful.

**Comment by dai-shi** [comment]
One thing to note is we prefer compact README. So, either to use `<details>` or a dedicated wiki page would be good. I don't mean to hide gotchas, and tbh I would like to describe such technical details. But, for the first look at README shouldn't take too long.



> But now I realized that re-render optimization is even more advance, which is a great selling point!

Yeah, currently it's not selling the point very well. btw, I've been working on this render optimization for years with [react-tracked](https://github.com/dai-shi/react-tracked), and valtio is a variant in this sense.
Also, wrote some blog posts like this: https://blog.axlight.com/posts/what-is-state-usage-tracking-a-novel-approach-to-intuitive-and-performant-api-with-react-hooks-and-proxy/

**Comment by dai-shi** [comment]
https://github.com/pmndrs/valtio/wiki/Some-gotchas is created.
Maybe, we should split pages once we have many items in the page.
